src/java.base/share/classes/jdk/internal/foreign/BufferStack.java

```diff
@@ -0,0 +1,188 @@
+            if (needsLock && !lock.tryLock()) {
+                // Rare: another virtual thread on the same carrier competed for acquisition.
+                return Arena.ofConfined();
+            }
+            if (!stack.canAllocate(size, byteAlignment)) {
+                if (needsLock) lock.unlock();
+                return Arena.ofConfined();
+            }
+            return new Frame(needsLock, size, byteAlignment);
+        }
+
+        static PerThread of(long size) {
+            final Arena arena = Arena.ofAuto();
+            final SlicingAllocator stack = new SlicingAllocator(arena.allocate(size));
+            return new PerThread(new ReentrantLock(), stack, arena);
+        }
+
+        private final class Frame implements Arena {
+            private final boolean locked;
+            private final long parentOffset;
+            private final long topOfStack;
+            private final Arena confinedArena = Arena.ofConfined();
+            private final SegmentAllocator frame;
+
+            @SuppressWarnings("restricted")
+            @ForceInline
+            public Frame(boolean locked, long byteSize, long byteAlignment) {
+                this.locked = locked;
+                parentOffset = stack.currentOffset();
+                MemorySegment frameSegment = stack.allocate(byteSize, byteAlignment);
+                topOfStack = stack.currentOffset();
+                frame = new SlicingAllocator(frameSegment.reinterpret(confinedArena, null));
+            }
+
+            @ForceInline
+            private void assertOrder() {
+                if (topOfStack != stack.currentOffset())
+                    throw new IllegalStateException("Out of order access: frame not top-of-stack");
+            }
+
+            @ForceInline
+            @Override
+            @SuppressWarnings("restricted")
+            public MemorySegment allocate(long byteSize, long byteAlignment) {
+                // Make sure we are on the right thread
+                if (((MemorySessionImpl) scope()).ownerThread() != Thread.currentThread()) {
+                    throw MemorySessionImpl.wrongThread();
+                }
+                return frame.allocate(byteSize, byteAlignment);
+            }
+
+            @ForceInline
+            @Override
+            public MemorySegment.Scope scope() {
+                return confinedArena.scope();
+            }
+
+            @ForceInline
+            @Override
+            public void close() {
+                assertOrder();
+                // the Arena::close method is called "early" as it checks thread
+                // confinement and crucially before any mutation of the internal
+                // state takes place.
+                confinedArena.close();
+                stack.resetTo(parentOffset);
+                if (locked) {
+                    lock.unlock();
+                }
+                Reference.reachabilityFence(arena);
```

Comment:
I'm not sure this is enough to keep the automatic arena alive. If the client lets the Frame arena go out of scope w/o calling close, then `arena` will become unreachable, but some segments created by the Frame arena might still be reachable. To be more correct, I think `Frame` should add a "close action" to its confined arena which keeps the outer automatic arena alive. This can be done, for instance, by passing a close action to the `reinterpret` call:

```
frame = new SlicingAllocator(frameSegment.reinterpret(confinedArena, () -> Reference.reachabilityFence(arena)));
```

The close action is installed in the `MemorySession` object of `confinedArena` -- which is then attached to all segments returned by `Frame` -- thus keeping the automatic arena alive.

Reply:
(if you agree with this analysis, perhaps adding an extra stress test, or tweaking one of the existing stress tests to check this could also be useful)

Reply:
I've thought about it a bit, and I think adding a cleanup action to the reinterpret call is indeed the only fool proof way to add a link back to the parent segment. MemorySegment is what gives a user access to the memory (almost like an access token), so the memory needs to be kept alive through the MS. MS has a reference to its scope, which can then reference the original scope through a cleanup action.